File: kernel/drivers/agnostic/pty.lua
local BaseTty = require("drivers.common.tty")

-- A PTY is a single character device with a full termios line discipline.
--
-- Data flow:
--   write_file(raw chars) → line discipline → read_queue → read_file (slave reads)
--   _write (echo / output processing) → output_buf → read_output (master reads)
--
-- Mount at a path like /dev/pty0. The writing process feeds raw input;
-- the reading process gets cooked lines.

---@class Pty: BaseTty
---@field output_buf table<integer, string>
local Pty = BaseTty:new()

---@param arch Arch
---@return Pty
function Pty:new(arch)
    local pty = BaseTty.new(self, arch)
    pty.output_buf = {}
    return pty
end

-- Echo and any output-processed data accumulate here.
-- Call read_output() to drain (e.g. for a terminal emulator to display).
function Pty:_write(data)
    table.insert(self.output_buf, data)
end

-- Read from the output side (echo + program output).
---@param length integer  0 = unlimited
---@return string?
function Pty:read_output(length)
    if #self.output_buf == 0 then return nil end
    local chunk = table.remove(self.output_buf, 1)
    if length and length > 0 and #chunk > length then
        table.insert(self.output_buf, 1, chunk:sub(length + 1))
        chunk = chunk:sub(1, length)
    end
    return chunk
end

return Pty

File: kernel/drivers/agnostic/pty.ts
import { BaseTty } from "../common/tty/init";

class Pty extends BaseTty {
	output_buf: string[] = [];

	private write(data: string) {
		this.output_buf.push(data);
	}

	read_output(length: integer) {
		if (this.output_buf.length === 0) { return undefined; }
		let chunk = this.output_buf.shift();
		if (chunk !== undefined && length > 0 && chunk.length >= length) {
			this.output_buf.push(string.sub(chunk, length + 1));
			chunk = string.sub(chunk, 1, length);
			return chunk;
		}
		return undefined;
	}

	write_raw(char: char): void { }
	echo_erase(n: integer): void { }
	set_cursor_pos(x: integer, y: integer): void { }
	get_cursor_pos(): [integer, integer] { return [0, 0]	}
	get_term_size(): [integer, integer] { return [0, 0] }
	clear_screen(): void { }
	clear_to_eol(): void { }
	clear_line(): void { }
	set_text_color(ansi: integer): void { }
	set_bg_color(ansi: integer): void { }
	reset_colors(): void { }
	set_cursor_visible(visible: boolean): void { }
}
